### DISCUSSION TITLE
I want to ask about Custodian policies for Infrastructure Code

### DISCUSSION THREAD

**Discussion by minhphuc2k21** [cloud-custodian/cloud-custodian#9720] [General]
Hi everyone,

I use c7n-left to test the governance of Terraform code and I use local cloud provider.
Example: 
main.tf
resource "vngcloud_vserver_server" "server" {
  count             = var.server_count
  project_id        = var.project_id
  name              = "vngcloud-server-1"
  encryption_volume = true
  attach_floating   = false
  flavor_id         = "123456"
  image_id          = "img-123456"
  network_id        = "net123456"
  root_disk_size    = 50
  root_disk_type_id = "volume-123456"
  security_group    = "sec-123445"
  subnet_id         = "sub-123456"
  action            = "stop"
}

resource "vngcloud_vserver_volume" "volume" {
  count          = var.server_count
  name           = "vngcloud-volume-1"
  size           = 50
  volume_type_id = "volume-123456"
  project_id     = "pro-1234567"
}
and i have policy.yaml

policies:
  - name: check-vserver-server
    resource: terraform.vngcloud_vserver_server
    filters:
      - type: value
        key: "flavor_id"
        value: "123456"
        op: eq

But when running c7n-left, the results are not in accordance with the policy I have set. This is result:
c7n-left run -p . -d .
Running 1 policies on 4 resources
check-vserver-server - terraform.vngcloud_vserver_server
  Failed
  File: ./main.tf:54-68
  54 resource "vngcloud_vserver_server" "server" {
  55   count             = var.server_count
  56   project_id        = var.project_id
  57   name              = "vngcloud-server-1"
  58   encryption_volume = true
  59   attach_floating   = false
  60   flavor_id         = "123456"
  61   image_id          = "img-123456"
  62   network_id        = "net123456"
  63   root_disk_size    = 50
  64   root_disk_type_id = "volume-123456"
  65   security_group    = "sec-123445"
  66   subnet_id         = "sub-123456"
  67   action            = "stop"
  68 }

Evaluation complete 0.01 seconds -> 1 Failures
                  Summary - By Policy
┏━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Severity ┃ Policy               ┃ Result            ┃
┡━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ unknown  │ check-vserver-server │ 1 failed 0 passed │
└──────────┴──────────────────────┴───────────────────┘
0 compliant of 4 total, 1 resources have 1 policy violations, 3 resources unevaluated
Can you help me explain this issue?
Thank you

**Comment by ajkerrigan** [comment]
Thanks for providing an example policy and module, I get 

```console
                  Summary - By Policy                  
┏━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Severity ┃ Policy               ┃ Result            ┃
┡━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ unknown  │ check-vserver-server │ 1 failed 0 passed │
└──────────┴──────────────────────┴───────────────────┘
0 compliant of 2 total, 1 resources have 1 policy violations, 1 resources unevaluated
```

Which seems to make sense given the policy. A "failure" is a resource that matches your policy, and there's one resource that matches.

I'm not sure where your output is getting the 4 total resources from - is there another Terraform file in the same directory perhaps? You can see the whole resource graph as c7n-left sees it with a command like:

```console
c7n-left dump -d .
```

And when I use that locally to look at the `main.tf` you provided, it shows the 2 resources I'd expect:

```json
{
  "input_vars": {},
  "graph": {
    "vngcloud_vserver_server": [
      {
        "__tfmeta": {
          "filename": "main.tf",
          "label": "vngcloud_vserver_server",
          "line_end": 15,
          "line_start": 1,
          "path": "vngcloud_vserver_server.server[0]",
          "type": "resource"
        },
        "action": "stop",
        "attach_floating": false,
        "count": null,
        "encryption_volume": true,
        "flavor_id": "123456",
        "id": "2b359f79-fe1c-49af-ab7a-74d717e9ec72",
        "image_id": "img-123456",
        "name": "vngcloud-server-1",
        "network_id": "net123456",
        "project_id": null,
        "root_disk_size": 50,
        "root_disk_type_id": "volume-123456",
        "security_group": "sec-123445",
        "subnet_id": "sub-123456"
      }
    ],
    "vngcloud_vserver_volume": [
      {
        "__tfmeta": {
          "filename": "main.tf",
          "label": "vngcloud_vserver_volume",
          "line_end": 23,
          "line_start": 17,
          "path": "vngcloud_vserver_volume.volume[0]",
          "type": "resource"
        },
        "count": null,
        "id": "12c02d2b-45c8-4101-970c-274f3e385de4",
        "name": "vngcloud-volume-1",
        "project_id": "pro-1234567",
        "size": 50,
        "volume_type_id": "volume-123456"
      }
    ]
  }
}
```

**Reply by minhphuc2k21** [reply]
Hi Ajkerrigan,

I'm glad to have received your response. 
I checked again and the results are exactly what you said. I would like to ask more about how this part of the policy will work. For example I want when running c7n-left run -p . -d. will scan and check main.tf if the flavor or capacity of the volume(size) is not as per my policy then it will Failure. 
Logically, I think if it matches the policy, the result will be Pass. Sorry if I misunderstood the meaning of the defined policy

**Reply by minhphuc2k21** [reply]
Regarding the action of sending notifications via webhook, I have a simple configuration when the job fails as follows: 
act:
      - Type: webhook
        url: "https://webhook.site/e94d80f9-7ce1-4ea5-8574-600fbf1fd676"
        method: POST
        body: |
          {
            "alert": "Check volume size",
            "message": "Policy-vserver-volume-size check passed.",
            "expected size": 50,
            "actual_size": "${resource.size}"
          }
        title:
          Content type: application/json
However, I did not receive any notification. I use webhook.site for quick testing
